### DISCUSSION TITLE
通过readable回调函数读取流quic_stream_read和直接读取流方式的不同

### DISCUSSION THREAD

**Discussion by sophiezou19** [Tencent/tquic#298] [Q&A (问题和回答)]
我发现不同方式读取流，会有以下现象：
1.通过readable回调函数读取流quic_stream_read，读完会向服务端回一个固定长度的包a（应该是ACK?）
2.通过直接读取流并调用内部处理quic_endpoint_process_connections，就不会发送。
想请问下：
现象1中回包a的作用，以及不回会不会导致服务端重传，或者其他影响？

**Comment by glime-ai** [comment]
现象1中，回包a的作用是发送ACK（确认包），用于确认已经成功接收到数据。这是QUIC协议的一部分，用于确保数据传输的可靠性。如果不发送ACK，服务端可能会认为数据包丢失，从而触发重传机制，导致不必要的网络开销和延迟。

现象2中，如果不发送ACK，确实可能会导致服务端重传数据包，因为服务端没有收到确认，认为数据可能没有成功到达。

为了确保数据传输的可靠性和效率，建议在读取流数据后，确保发送ACK包。

This comment was generated by [Glime](https://glimelab.ai/?utm_source=44520fce49ba43798485a628d6928e22&utm_medium=github).

**Comment by iyangsj** [comment]
>1.通过readable回调函数读取流quic_stream_read，读完会向服务端回一个固定长度的包a（应该是ACK?）

ACK的回复与应用层是否已读取缓存的数据无关。

> 2.通过直接读取流并调用内部处理quic_endpoint_process_connections，就不会发送。

同上，一种可能是ACK已提前发送。

这两种情况可以提供下debug级别的日志看看

**Reply by sophiezou19** [reply]
再请问下，readable事件也是水平触发吗，readable回调中我先只读了部分流，然后quic_stream_wantread置为false，完成后，在外部方法中将quic_stream_wantread再置为true并调内部处理，发现并没有立即发起可读回调事件来读取剩余流。

**Reply by iyangsj** [reply]
> 完成后，在外部方法中将quic_stream_wantread再置为true并调内部处理，发现并没有立即发起可读回调事件来读取剩余流。

这种情况下，在执行quic_stream_wantread后，立即发起一次quic_stream_read
